# Big Data Final Project -- CDC Data Set With All Year 
Lu Zhang, Yuqian Wang, Yue Chen, Lu Wang, Jihan Gao

# Step 1: Load Data

In [3]:
#import findspark
import pyspark
from pyspark.context import SparkContext
from pyspark.sql.session import SparkSession
from pyspark import mllib
from pyspark import ml
#findspark.init("/opt/spark-2.4.0")
#sc = SparkContext('local')
#spark = SparkSession(sc)
from pyspark.mllib.tree import DecisionTree, DecisionTreeModel
from pyspark.mllib.util import MLUtils
from pyspark.mllib.regression import LabeledPoint
from pyspark import SparkConf, SparkContext
from numpy import array
from pyspark.sql.types import IntegerType
from pyspark.sql.types import DoubleType
from pyspark.sql import *
from pyspark.sql import SparkSession
from pyspark.ml.feature import StringIndexer
from pyspark.sql.functions import udf
from pyspark.sql import Row
from pyspark.ml.linalg import Vectors
from pyspark.ml.classification import MultilayerPerceptronClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.sql.types import IntegerType
from pyspark.sql.types import DoubleType
from pyspark.ml.feature import *
import numpy as np
import pandas as pd
from pyspark.ml.feature import Tokenizer, RegexTokenizer
from pyspark.sql.functions import col, udf
from pyspark.sql.types import IntegerType
from pyspark.ml.feature import ChiSqSelector
from pyspark.ml.linalg import Vectors

In [4]:
# File location and type
file_location = "/FileStore/tables/2015_data.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "true"
delimiter = ","

# Upload all dataset from year 2005 - 2015
df_2015 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

df_2014 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2014_data.csv")

df_2013 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2013_data.csv")

df_2012 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2012_data.csv")

df_2011 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2011_data.csv")

df_2010 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2010_data.csv")

df_2009 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2009_data.csv")

df_2008 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2008_data.csv")

df_2007 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2007_data.csv")

df_2006 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2006_data.csv")

df_2005 = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load("/FileStore/tables/2005_data.csv")


In [5]:
# Add year column to each dataframe
num_2005 = df_2005.count()
addyear = udf(lambda x: 2005, returnType=IntegerType()) #Define UDF function
df_2005 = df_2005.withColumn('year',addyear('detail_age'))

num_2006 = df_2006.count()
addyear = udf(lambda x: 2006, returnType=IntegerType()) #Define UDF function
df_2006 = df_2006.withColumn('year',addyear('detail_age'))

num_2007 = df_2007.count()
addyear = udf(lambda x: 2007, returnType=IntegerType()) #Define UDF function
df_2007 = df_2007.withColumn('year',addyear('detail_age'))

num_2008 = df_2008.count()
addyear = udf(lambda x: 2008, returnType=IntegerType()) #Define UDF function
df_2008 = df_2008.withColumn('year',addyear('detail_age'))

num_2009 = df_2009.count()
addyear = udf(lambda x: 2009, returnType=IntegerType()) #Define UDF function
df_2009 = df_2005.withColumn('year',addyear('detail_age'))

num_2010 = df_2010.count()
addyear = udf(lambda x: 2010, returnType=IntegerType()) #Define UDF function
df_2010 = df_2010.withColumn('year',addyear('detail_age'))

num_2011 = df_2011.count()
addyear = udf(lambda x: 2011, returnType=IntegerType()) #Define UDF function
df_2011 = df_2011.withColumn('year',addyear('detail_age'))

num_2012 = df_2012.count()
addyear = udf(lambda x: 2012, returnType=IntegerType()) #Define UDF function
df_2012 = df_2012.withColumn('year',addyear('detail_age'))

num_2013 = df_2013.count()
addyear = udf(lambda x: 2013, returnType=IntegerType()) #Define UDF function
df_2013 = df_2013.withColumn('year',addyear('detail_age'))

num_2014 = df_2014.count()
addyear = udf(lambda x: 2014, returnType=IntegerType()) #Define UDF function
df_2014 = df_2014.withColumn('year',addyear('detail_age'))

num_2015 = df_2015.count()
addyear = udf(lambda x: 2015, returnType=IntegerType()) #Define UDF function
df_2015 = df_2015.withColumn('year',addyear('detail_age'))


In [6]:
#combine the 10 dataframe into one dataframe as df. 
df = df_2005.union(df_2006).union(df_2007).union(df_2008).union(df_2009).union(df_2010).union(df_2011).union(df_2012).union(df_2013).union(df_2014).union(df_2015)
display(df)

resident_status education_1989_revision education_2003_revision education_reporting_flag month_of_death sex detail_age_type detail_age age_substitution_flag age_recode_52 age_recode_27 age_recode_12 infant_age_recode_22 place_of_death_and_decedents_status marital_status day_of_week_of_death current_data_year injury_at_work manner_of_death method_of_disposition autopsy activity_code place_of_injury_for_causes_w00_y34_except_y06_and_y07_ icd_code_10th_revision 358_cause_recode 113_cause_recode 130_infant_cause_recode 39_cause_recode number_of_entity_axis_conditions entity_condition_1 entity_condition_2 entity_condition_3 entity_condition_4 entity_condition_5 entity_condition_6 entity_condition_7 entity_condition_8 entity_condition_9 entity_condition_10 entity_condition_11 entity_condition_12 entity_condition_13 entity_condition_14 entity_condition_15 entity_condition_16 entity_condition_17 entity_condition_18 entity_condition_19 entity_condition_20 number_of_record_axis_conditions record_condition_1 record_condition_2 record_condition_3 record_condition_4 record_condition_5 record_condition_6 record_condition_7 record_condition_8 record_condition_9 record_condition_10 record_condition_11 record_condition_12 record_condition_13 record_condition_14 record_condition_15 record_condition_16 record_condition_17 record_condition_18 record_condition_19 record_condition_20 race bridged_race_flag race_imputation_flag race_recode_3 race_recode_5 hispanic_origin hispanic_originrace_recode year 1 11 null 0 01 F 1 045 null 35 15 07 null 1 M 2 2005 U 7 U N null null C439 098 028 null 15 01 11C439 null null null null null null null null null null null null null null null null null null null 01 C439 null null null null null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 13 null 0 01 M 1 061 null 38 18 08 null 1 D 7 2005 U 7 U N null null J439 266 084 null 28 01 11J439 null null null null null null null null null null null null null null null null null null null 01 J439 null null null null null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 12 null 0 01 F 1 079 null 41 21 10 null 6 D 1 2005 U 7 U N null null I698 239 070 null 24 05 11R628 21I698 61J449 62M199 63R568 null null null null null null null null null null null null null null null 05 I698 J449 M199 R568 R628 null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 12 null 0 01 M 1 050 null 36 16 07 null 1 S 4 2005 U 7 U N null null E119 159 046 null 16 04 11I469 61E119 62I500 63K862 null null null null null null null null null null null null null null null null 04 E119 I469 I500 K862 null null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 14 null 0 01 F 1 068 null 39 19 09 null 1 M 2 2005 U 7 U N null null C349 093 027 null 08 03 11C349 61F179 62J449 null null null null null null null null null null null null null null null null null 03 C349 F179 J449 null null null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 03 null 0 01 F 1 089 null 43 23 11 null 6 W 7 2005 U 7 U N null null I679 239 070 null 24 03 11N19 21I679 61F03 null null null null null null null null null null null null null null null null null 03 F03 I679 N19 null null null null null null null null null null null null null null null null null 03 null null 2 3 100 8 2005 1 12 null 0 01 F 1 068 null 39 19 09 null 1 D 7 2005 U 7 U N null null J439 266 084 null 28 01 11J439 null null null null null null null null null null null null null null null null null null null 01 J439 null null null null null null null null null null null null null null null null null null null 01 null null 1 1 100 6 2005 1 12 null 0 01 M 1 061 null 38 18 08 null 1 S 6 2005 U 7 U N null null J449 267 086 null 28 03 11I279 21J449 61F102 null null null null null null null null null null null null null 

# Step 2: Feature Selection

In [8]:
#Filter out column with no data 
names_test = df.columns
for name in names_test:
    print(name + "  ")
    print(df.select(name).dropna().count())
    
names = df.columns
names.remove("method_of_disposition")
names.append("method_of_disposition")
to_be_removed = ["race_imputation_flag", "bridged_race_flag", "record_condition_20", \
                 "record_condition_19", "record_condition_18", "record_condition_17", \
                 "record_condition_16", "record_condition_15", "record_condition_14", \
                 "record_condition_13", "record_condition_12", "record_condition_11", \
                 "record_condition_10", "record_condition_9", "record_condition_8", \
                 "record_condition_7", "record_condition_6", "record_condition_5", \
                 "record_condition_4", "record_condition_3", "record_condition_2", \
                 "record_condition_1", "130_infant_cause_recode", \
                 "education_1989_revision", "age_substitution_flag", "infant_age_recode_22", \
                 "activity_code", "place_of_injury_for_causes_w00_y34_except_y06_and_y07_", \
                 "entity_condition_1", "entity_condition_2", "entity_condition_3", "entity_condition_4", \
                 "entity_condition_5", "entity_condition_6","entity_condition_7", "entity_condition_8", \
                 "entity_condition_9", "entity_condition_10", "entity_condition_11", "entity_condition_12", \
                 "entity_condition_13", "entity_condition_14","entity_condition_15",  "entity_condition_16", \
                 "entity_condition_17", "entity_condition_18","entity_condition_19", "entity_condition_20"]

#Remove columns of df which contains no data
for removed in to_be_removed:
        names.remove(removed)

df_feature_select = df.select(names)
df_feature_select = df_feature_select.dropna()
df_feature_select.show()

resident_status 
27731960
education_1989_revision 
8814633
education_2003_revision 
18917327
education_reporting_flag 
27731960
month_of_death 
27731960
sex 
27731960
detail_age_type 
27731960
detail_age 
27731960
age_substitution_flag 
975
age_recode_52 
27731960
age_recode_27 
27731960
age_recode_12 
27731960
infant_age_recode_22 
285690
place_of_death_and_decedents_status 
27731960
marital_status 
27731960
day_of_week_of_death 
27731960
current_data_year 
27731960
injury_at_work 
27731960
manner_of_death 
21863725
method_of_disposition 
27731960
autopsy 
27731960
activity_code 
2036916
place_of_injury_for_causes_w00_y34_except_y06_and_y07_ 
1928869
icd_code_10th_revision 
27731960
358_cause_recode 
27731960
113_cause_recode 
27731960
130_infant_cause_recode 
285690
39_cause_recode 
27731960
number_of_entity_axis_conditions 
27731960
entity_condition_1 
27731960
entity_condition_2 
21491514
entity_condition_3 
14891671
entity_condition_4 
8895389
entity_condition_5 
4850109
entity_condition_6 
2531032
entity_condition_7 
1276857
entity_condition_8 
630079
entity_condition_9 
308568
entity_condition_10 
147382
entity_condition_11 
69099
entity_condition_12 
31116
entity_condition_13 
12893
entity_condition_14 
5458
entity_condition_15 
656
entity_condition_16 
177
entity_condition_17 
61
entity_condition_18 
22
entity_condition_19 
8
entity_condition_20 
0
number_of_record_axis_conditions 
27731960
record_condition_1 
27731960
record_condition_2 
21107395
record_condition_3 
14116097
record_condition_4 
8232094
record_condition_5 
4317769
record_condition_6 
2139791
record_condition_7 
1018220
record_condition_8 
471721
record_condition_9 
214479
record_condition_10 
94103
record_condition_11 
39980
record_condition_12 
15760
record_condition_13 
5340
record_condition_14 
1508
record_condition_15 
234
record_condition_16 
56
record_condition_17 
21
record_condition_18 
7
record_condition_19 
3
record_condition_20 
0
race 
27731960
bridged_race_flag 
83317
race_imputation_flag 
179468
race_recode_3 
27731960
race_recode_5 
27731960
hispanic_origin 
27731960
hispanic_originrace_recode 
27731960
year 
27731960
+---------------+-----------------------+------------------------+--------------+---+---------------+----------+-------------+-------------+-------------+-----------------------------------+--------------+--------------------+-----------------+--------------+---------------+-------+----------------------+----------------+----------------+---------------+--------------------------------+--------------------------------+----+-------------+-------------+---------------+--------------------------+----+---------------------+
resident_status|education_2003_revision|education_reporting_flag|month_of_death|sex|detail_age_type|detail_age|age_recode_52|age_recode_27|age_recode_12|place_of_death_and_decedents_status|marital_status|day_of_week_of_death|current_data_year|injury_at_work|manner_of_death|autopsy|icd_code_10th_revision|358_cause_recode|113_cause_recode|39_cause_recode|number_of_entity_axis_conditions|number_of_record_axis_conditions|race|race_recode_3|race_recode_5|hispanic_origin|hispanic_originrace_recode|year|method_of_disposition|
+---------------+-----------------------+------------------------+--------------+---+---------------+----------+-------------+-------------+-------------+-----------------------------------+--------------+--------------------+-----------------+--------------+---------------+-------+----------------------+----------------+----------------+---------------+--------------------------------+--------------------------------+----+-------------+-------------+---------------+--------------------------+----+---------------------+
 1| 3| 1| 01| M| 1| 057| 37| 17| 08| 4| M| 2| 2005| N| 2| N| X72| 429| 125| 40| 04| 03| 01| 1| 1| 100| 6|2005| C|
 1| 1| 1| 01| F| 1| 077| 41| 21| 10| 6| W| 3| 2005| U| 7| N| I694| 239| 070| 24| 02| 02| 01| 1| 1| 210| 1|2005| B|
 2| 3| 1| 01| M| 1| 095| 45| 25| 11| 6| S| 4| 20

In [9]:
#TRANSFORM STRING TO FLOAT
df_feature_select_dup = df_feature_select
for name in names:
    indexer = StringIndexer(inputCol=name, outputCol=name + "_index")
    df_feature_select_dup = indexer.fit(df_feature_select_dup).transform(df_feature_select_dup)

for name in names:
    df_feature_select_dup = df_feature_select_dup.drop(name)

df_feature_select_dup.columns

def transData(data):
    return data.rdd.map(lambda r: [r[-1], Vectors.dense(r[:-1])]). \
           toDF(['labelString','features'])
    
df_feature_select_trans=transData(df_feature_select_dup)
df_feature_select_trans.show()


#### Use ChiSqSelector to select top 10 features

In [11]:

#Create ChiSqSelector to list top 10 independent features
selector = ChiSqSelector(numTopFeatures=10, featuresCol="features",
                         outputCol="selectedFeatures", labelCol="labelString")
#Fit data to selector
result = selector.fit(df_feature_select_trans).transform(df_feature_select_trans)

def extract(row):
    return (row.labelString, ) + tuple(float(x) for x in row.features.values) + tuple(float(x) for x in row.selectedFeatures.values)
result_featuers_df = result.rdd.map(extract).toDF(["labelString"])

#Convert to Panda dataframe
pd_df = result_featuers_df.toPandas()

#Compare values of original columns with values of selected top 10 columns
for i in range(2,30):
  for j in range(30,40):
    feature_col = "_" + str(i)
    selected_col = "_" + str(j)
    if(pd_df[feature_col].equals(pd_df[selected_col])):
      print(names[i])


# Step 3: Select 'Resident Status', 'Education, Sex', 'Age_recode_12', 'Marital_status', 'Month_of_death', 'Place_of_death' as independent features for building model

In [13]:
df = df['resident_status', 'education_2003_revision', 'sex', 'age_recode_12', 'marital_status', 'month_of_death','place_of_death_and_decedents_status','method_of_disposition']

rs = udf(lambda x: 1 if x == '2' else 2 if x == '3' else 3 if x == '4' else 0)
edu = udf(lambda x: 1 if x == '2' else 2 if x == '3' else 3 if x == '4' else 4 if x == '5' else 5 if x == '6' else 6 if x == '7' else 7 if x == '8' else 0)
se = udf(lambda x: 1 if x == 'F' else 0)
recode12 = udf(lambda x: 1 if x == '2' else 2 if x == '3' else 3 if x == '4' else 4 if x == '5' else 5 if x == '6' else 6 if x == '7' else 7 if x == '8' else 8 if x == '9' else 9 if x == '10' else 10 if x == '11' else 0)
mar = udf(lambda x: 1 if x == 'S' else 2 if x == 'W' else 3 if x == 'D' else 4 if x == 'M' else 0)
mon = udf(lambda x: 1 if x == '02' else 2 if x == '03' else 3 if x == '04' else 4 if x == '05' else 5 if x == '06' else 6 if x == '07' else 7 if x == '08' else 8 if x == '09' else 9 if x == '10' else 10 if x == '11' else 11 if x == '12' else 0)
pla = udf(lambda x: 1 if x == '2' else 2 if x == '3' else 3 if x == '4' else 4 if x == '5' else 5 if x == '6' else 6 if x == '7' else 7 if x == '9' else 0)
method = udf(lambda x: 1 if x == 'B' else 2 if x == 'C' else 0 )

#convert all Character category to integer category & Convert Column type to Double and 
df = df.withColumn("resident_status",rs(df['resident_status']).cast(DoubleType()))
df = df.withColumn("education_2003_revision",edu(df['education_2003_revision']).cast(DoubleType()))
df = df.withColumn("sex",se(df['sex']).cast(DoubleType()))
df = df.withColumn("age_recode_12",recode12(df['age_recode_12']).cast(DoubleType()))
df = df.withColumn("marital_status",mar(df['marital_status']).cast(DoubleType()))
df = df.withColumn("month_of_death",mon(df['month_of_death']).cast(DoubleType()))
df = df.withColumn("place_of_death_and_decedents_status",pla(df['place_of_death_and_decedents_status']).cast(DoubleType()))
df = df.withColumn("method_of_disposition",method(df['method_of_disposition']).cast(DoubleType()))
df = df.dropna()
df.show()


+---------------+-----------------------+---+-------------+--------------+--------------+-----------------------------------+---------------------+
resident_status|education_2003_revision|sex|age_recode_12|marital_status|month_of_death|place_of_death_and_decedents_status|method_of_disposition|
+---------------+-----------------------+---+-------------+--------------+--------------+-----------------------------------+---------------------+
 0.0| 0.0|1.0| 0.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 0.0| 3.0| 0.0| 0.0| 0.0|
 0.0| 0.0|1.0| 9.0| 3.0| 0.0| 5.0| 0.0|
 0.0| 0.0|0.0| 0.0| 1.0| 0.0| 0.0| 0.0|
 0.0| 0.0|1.0| 0.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|1.0| 10.0| 2.0| 0.0| 5.0| 0.0|
 0.0| 0.0|1.0| 0.0| 3.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 0.0| 1.0| 0.0| 0.0| 0.0|
 0.0| 0.0|1.0| 0.0| 3.0| 0.0| 0.0| 0.0|
 2.0| 0.0|1.0| 10.0| 2.0| 0.0| 6.0| 0.0|
 0.0| 0.0|1.0| 10.0| 2.0| 0.0| 5.0| 0.0|
 0.0| 0.0|0.0| 9.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 9.0| 3.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 0.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 9.0| 2.0| 0.0| 0.0| 0.0|
 0.0| 0.0|1.0| 0.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 9.0| 4.0| 0.0| 5.0| 0.0|
 0.0| 0.0|0.0| 0.0| 4.0| 0.0| 0.0| 0.0|
 0.0| 0.0|0.0| 0.0| 3.0| 0.0| 6.0| 0.0|
 0.0| 0.0|1.0| 10.0| 2.0| 0.0| 5.0| 0.0|
+---------------+-----------------------+---+-------------+--------------+--------------+-----------------------------------+---------------------+
only showing top 20 rows

# Model 1:Decision Tree Model

In [15]:
(trainingData1, testData1) = df.randomSplit([0.7, 0.3])

#Create labeled point to fit into Decision tree model
def createdLabeledPoints(fields):
  resident_status = fields[0]
  education_2003_revision = fields[1]
  sex = fields[2]
  age_recode_12 = fields[3]
  marital_status = fields[4]
  month_of_death = fields[5]
  place_of_death_and_decedents_status = fields[6]
  method_of_disposition = fields[7]
  
  return LabeledPoint(method_of_disposition, array([resident_status, education_2003_revision, sex, age_recode_12, marital_status, month_of_death, place_of_death_and_decedents_status]))

#Convert training data to rdd inorder to fit into decision model
rdd = trainingData1.rdd
trainingRdd = rdd.map(createdLabeledPoints)

#create a Decision Tree model with max Depth 7 and Maxbins 79
model = DecisionTree.trainClassifier(trainingRdd, numClasses = 3, 
                                     categoricalFeaturesInfo = {0:4, 1:9, 2:2, 3:12, 4:5, 5:12, 6:8}, 
                                     impurity = 'gini', maxDepth = 7, maxBins = 79)

print(model.toDebugString())


DecisionTreeModel classifier of depth 7 with 255 nodes
 If (feature 1 in {0.0})
 If (feature 6 in {0.0,5.0,1.0,6.0,2.0,7.0,3.0})
 If (feature 4 in {0.0,1.0,2.0})
 If (feature 4 in {1.0,0.0})
 If (feature 6 in {5.0,6.0,2.0,7.0,3.0})
 If (feature 6 in {7.0})
 If (feature 4 in {0.0})
 Predict: 0.0
 Else (feature 4 not in {0.0})
 Predict: 0.0
 Else (feature 6 not in {7.0})
 If (feature 4 in {0.0})
 Predict: 1.0
 Else (feature 4 not in {0.0})
 Predict: 0.0
 Else (feature 6 not in {5.0,6.0,2.0,7.0,3.0})
 If (feature 4 in {1.0})
 If (feature 3 in {10.0,9.0})
 Predict: 0.0
 Else (feature 3 not in {10.0,9.0})
 Predict: 0.0
 Else (feature 4 not in {1.0})
 If (feature 3 in {10.0,9.0})
 Predict: 1.0
 Else (feature 3 not in {10.0,9.0})
 Predict: 1.0
 Else (feature 4 not in {1.0,0.0})
 If (feature 3 in {0.0,9.0})
 If (feature 6 in {6.0,7.0,2.0})
 If (feature 2 in {0.0})
 Predict: 0.0
 Else (feature 2 not in {0.0})
 Predict: 0.0
 Else (feature 6 not in {6.0,7.0,2.0})
 If (feature 3 in {0.0})
 Predict: 0.0
 Else (feature 3 not in {0.0})
 Predict: 0.0
 Else (feature 3 not in {0.0,9.0})
 If (feature 6 in {6.0,2.0,7.0})
 If (feature 6 in {6.0})
 Predict: 0.0
 Else (feature 6 not in {6.0})
 Predict: 0.0
 Else (feature 6 not in {6.0,2.0,7.0})
 If (feature 6 in {5.0,0.0})
 Predict: 0.0
 Else (feature 6 not in {5.0,0.0})
 Predict: 0.0
 Else (feature 4 not in {0.0,1.0,2.0})
 If (feature 3 in {0.0})
 If (feature 0 in {2.0,1.0})
 If (feature 6 in {7.0,2.0,6.0})
 If (feature 4 in {3.0})
 Predict: 0.0
 Else (feature 4 not in {3.0})
 Predict: 0.0
 Else (feature 6 not in {7.0,2.0,6.0})
 If (feature 0 in {1.0})
 Predict: 0.0
 Else (feature 0 not in {1.0})
 Predict: 0.0
 Else (feature 0 not in {2.0,1.0})
 If (feature 6 in {5.0,1.0,6.0,2.0,7.0,3.0})
 If (feature 6 in {2.0,7.0,6.0})
 Predict: 0.0
 Else (feature 6 not in {2.0,7.0,6.0})
 Predict: 0.0
 Else (feature 6 not in {5.0,1.0,6.0,2.0,7.0,3.0})
 If (feature 4 in {3.0})
 Predict: 0.0
 Else (feature 4 not in {3.0})
 Predict: 0.0
 Else (feature 3 not in {0.0})
 If (feature 6 in {6.0,7.0,2.0})
 If (feature 4 in {3.0})
 If (feature 2 in {0.0})
 Predict: 0.0
 Else (feature 2 not in {0.0})
 Predict: 0.0
 Else (feature 4 not in {3.0})
 If (feature 6 in {6.0,7.0})
 Predict: 0.0
 Else (feature 6 not in {6.0,7.0})
 Predict: 0.0
 Else (feature 6 not in {6.0,7.0,2.0})
 If (feature 3 in {9.0})
 If (feature 4 in {3.0})
 Predict: 0.0
 Else (feature 4 not in {3.0})
 Predict: 0.0
 Else (feature 3 not in {9.0})
 If (feature 4 in {3.0})
 Predict: 0.0
 Else (feature 4 not in {3.0})
 Predict: 0.0
 Else (feature 6 not in {0.0,5.0,1.0,6.0,2.0,7.0,3.0})
 If (feature 4 in {0.0,1.0,3.0})
 If (feature 4 in {3.0,0.0})
 If (feature 3 in {0.0})
 If (feature 0 in {2.0,1.0})
 If (feature 0 in {2.0})
 Predict: 2.0
 Else (feature 0 not in {2.0})
 Predict: 2.0
 Else (feature 0 not in {2.0,1.0})
 If (feature 0 in {0.0})
 Predict: 2.0
 Else (feature 0 not in {0.0})
 Predict: 2.0
 Else (feature 3 not in {0.0})
 If (feature 0 in {2.0,1.0})
 If (feature 0 in {1.0})
 Predict: 1.0
 Else (feature 0 not in {1.0})
 Predict: 1.0
 Else (feature 0 not in {2.0,1.0})
 If (feature 2 in {0.0})
 Predict: 2.0
 Else (feature 2 not in {0.0})
 Predict: 2.0
 Else (feature 4 not in {3.0,0.0})
 If (feature 3 in {10.0})
 If (feature 5 in {0.0,10.0,6.0,2.0,3.0})
 If (feature 0 in {0.0})
 Predict: 1.0
 Else (feature 0 not in {0.0})
 Predict: 1.0
 Else (feature 5 not in {0.0,10.0,6.0,2.0,3.0})
 If (feature 0 in {1.0,0.0})
 Predict: 1.0
 Else (feature 0 not in {1.0,0.0})
 Predict: 1.0
 Else (feature 3 not in {10.0})
 If (feature 0 in {1.0})
 If (feature 2 in {0.0})
 Predict: 1.0
 Else (feature 2 not in {0.0})
 Predict: 1.0
 Else (feature 0 not in {1.0})
 If (feature 2 in {0.0})
 Predict: 2.0
 Else (feature 2 not in {0.0})
 Predict: 1.0
 Else (feature 4 not in {0.0,1.0,3.0})
 If (feature 3 in {10.0,9.0})
 If (feature 0 in {1.0})
 If (feature 3 in {10.0})
 If (feature 5 in {0.0,5.0,6.0,7.0,11.0})
 Predict: 1.0
 Else (feature 5 not in {0.0,5.0,6.0,7.0,11.0})
 Predict: 1.0
 

####Test Decision Tree

In [17]:
testrdd = testData1.rdd
testingData = testrdd.map(createdLabeledPoints)
predictions = model.predict(testingData.map(lambda x: x.features))
labelsAndPredictions = testingData.map(lambda lp: lp.label).zip(predictions)
testErr = labelsAndPredictions.filter(lambda vp: vp[0] != vp[1]).count() /float(testingData.count())
print('Testing Error = ' + str(testErr))
print('Learned classification tree model:')
print(model)

Testing Error = 0.3512157714873731
Learned classification tree model:
DecisionTreeModel classifier of depth 7 with 255 nodes

###### Decision Tree model:  Accuracy of testing data: 64.88%

# Model 2: Multilayer Neuron Network

In [20]:
def transData(data):
    return data.rdd.map(lambda r: [r[-1], Vectors.dense(r[:-1])]).\
           toDF(['label','features'])
    
df2= transData(df)

(trainingData2, testData2) = df2.randomSplit([0.7, 0.3])
trainingData2=trainingData2["label","features"]
trainingData2.first()

# Specify layers for the neural network:
# Input layer: size 7 (number of input features)
# 8 intermediate layer: size 10,10,15,3,7,8,12,20
# Output layer: size 3 (number of categories on predictor variable)
layers = [7, 10, 10, 15, 3, 7, 8, 12, 20, 3]

# Create the trainer with 200 times iteration
FNN = MultilayerPerceptronClassifier(maxIter=100, layers=layers, blockSize=128, seed=1234)

# Fit training data into model
model = FNN.fit(trainingData2)

testData2=testData2["label","features"]
testData2.first()

Out[ 9 ]: Row(label=0.0, features=DenseVector([0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]))

####Test Multilayer Neuron Network

In [22]:
# compute accuracy on the test set
result = model.transform(testData2)
predictionAndLabels = result.select("prediction", "label")
evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
print("Test set accuracy = " + str(evaluator.evaluate(predictionAndLabels)))

Test set accuracy = 0.6157357887496846

###### Multi Layer Neuron Network:  Accuracy of testing data: 61.57%

# Model 3: Multinomial Logistic Regression

In [25]:
from pyspark.ml.classification import LogisticRegression

df_multi_log = df2["label","features"]
df_multi_log.show()

# Load training data
(trainingData3, testData3) = df_multi_log.randomSplit([0.7, 0.3])

# Build multinomial logistic regression
lr = LogisticRegression(maxIter=100, regParam=0.3, elasticNetParam=0.8)

# Fit the model
lrModel = lr.fit(trainingData3)

# Print the coefficients and intercept for multinomial logistic regression
print("Coefficients: \n" + str(lrModel.coefficientMatrix))
print("Intercept: " + str(lrModel.interceptVector))

trainingSummary = lrModel.summary

+-----+--------------------+
label| features|
+-----+--------------------+
 0.0|[0.0,0.0,1.0,0.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,1.0,9.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,1.0,0.0,...|
 0.0|[0.0,0.0,1.0,10.0...|
 0.0|[0.0,0.0,1.0,0.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,1.0,0.0,...|
 0.0|[2.0,0.0,1.0,10.0...|
 0.0|[0.0,0.0,1.0,10.0...|
 0.0|[0.0,0.0,0.0,9.0,...|
 0.0|[0.0,0.0,0.0,9.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,0.0,9.0,...|
 0.0|[0.0,0.0,1.0,0.0,...|
 0.0|[0.0,0.0,0.0,9.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,0.0,0.0,...|
 0.0|[0.0,0.0,1.0,10.0...|
+-----+--------------------+
only showing top 20 rows

Coefficients: 
3 X 7 CSRMatrix
(0,1) -0.0606
Intercept: [0.14843798831,0.0372146321632,-0.185652620473]

####Test Multi Nominal Regression model

In [27]:
# Obtain the objective per iteration
objectiveHistory = trainingSummary.objectiveHistory
print("objectiveHistory:")
for objective in objectiveHistory:
    print(objective)

# for multiclass, we can inspect metrics on a per-label basis
print("False positive rate by label:")
for i, rate in enumerate(trainingSummary.falsePositiveRateByLabel):
    print("label %d: %s" % (i, rate))

print("True positive rate by label:")
for i, rate in enumerate(trainingSummary.truePositiveRateByLabel):
    print("label %d: %s" % (i, rate))

print("Precision by label:")
for i, prec in enumerate(trainingSummary.precisionByLabel):
    print("label %d: %s" % (i, prec))

print("Recall by label:")
for i, rec in enumerate(trainingSummary.recallByLabel):
    print("label %d: %s" % (i, rec))

print("F-measure by label:")
for i, f in enumerate(trainingSummary.fMeasureByLabel()):
    print("label %d: %s" % (i, f))

accuracy = trainingSummary.accuracy
falsePositiveRate = trainingSummary.weightedFalsePositiveRate
truePositiveRate = trainingSummary.weightedTruePositiveRate
fMeasure = trainingSummary.weightedFMeasure()
precision = trainingSummary.weightedPrecision
recall = trainingSummary.weightedRecall
print("Accuracy: %s\nFPR: %s\nTPR: %s\nF-measure: %s\nPrecision: %s\nRecall: %s"
      % (accuracy, falsePositiveRate, truePositiveRate, fMeasure, precision, recall))

objectiveHistory:
1.0929434456167948
1.0926827780990582
1.0922300390460757
1.0919613781546047
1.0918267033894666
1.0916998365870074
1.091612847051117
1.0914679564481706
1.091367327789285
1.091348849729078
1.0913462653262507
1.091345941427183
1.0913457831653415
1.0913456427447554
1.0913455300791395
1.0913454476815572
1.091345385884754
1.0913453426171544
1.0913452098370344
1.0913452058699722
1.0913452038501996
1.091345198845416
1.0913451987044007
1.0913451986524327
1.0913451985719984
1.0913451985340392
1.091345198502512
1.0913451984761344
1.091345198447878
1.0913451984460405
False positive rate by label:
label 0: 0.2677133383342013
label 1: 0.3924036008518281
label 2: 0.0
True positive rate by label:
label 0: 0.9189601122923342
label 1: 0.6872956054824876
label 2: 0.0
Precision by label:
label 0: 0.6599944755732565
label 1: 0.4906532438126333
label 2: 0.0
Recall by label:
label 0: 0.9189601122923342
label 1: 0.6872956054824876
label 2: 0.0
F-measure by label:
label 0: 0.7682407107160568
label 1: 0.5725610555839391
label 2: 0.0
Accuracy: 0.575824972280066
FPR: 0.23594191146609544
TPR: 0.575824972280066
F-measure: 0.4806700084694489
Precision: 0.41250535153052637
Recall: 0.575824972280066

In [28]:
predicted_result = lrModel.transform(testData3)
predicted_result
evaluator = MulticlassClassificationEvaluator(
    labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluator.evaluate(predicted_result)
print(accuracy)

0.576208538610548

###### Multinomial Regression model: Accuracy: 57.62%

###### PS: The chisqselector is supposed to display top 10 features. However, the cluster failed on stage process when we run last time.